flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushFilterInCalcIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,139 @@
+        // we can not push filter twice
+        return canPushdownFilter(tableSourceTable);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        Calc calc = call.rel(0);
+        FlinkLogicalTableSourceScan scan = call.rel(1);
+        TableSourceTable table = scan.getTable().unwrap(TableSourceTable.class);
+        pushFilterIntoScan(call, calc, scan, table);
+    }
+
+    private void pushFilterIntoScan(
+            RelOptRuleCall call,
+            Calc calc,
+            FlinkLogicalTableSourceScan scan,
+            FlinkPreparingTableBase relOptTable) {
+
+        RexProgram originProgram = calc.getProgram();
+
+        RelBuilder relBuilder = call.builder();
+        Tuple2<RexNode[], RexNode[]> extractedPredicates =
+                extractPredicates(
+                        originProgram.getInputRowType().getFieldNames().toArray(new String[0]),
+                        originProgram.expandLocalRef(originProgram.getCondition()),
+                        scan,
+                        relBuilder.getRexBuilder());
+
+        RexNode[] convertiblePredicates = extractedPredicates._1;
+        if (convertiblePredicates.length == 0) {
+            // no condition can be translated to expression
+            return;
+        }
+
+        Tuple2<SupportsFilterPushDown.Result, TableSourceTable> pushdownResultWithScan =
+                createTableScanAfterPushdown(
+                        convertiblePredicates,
+                        relOptTable.unwrap(TableSourceTable.class),
+                        scan,
+                        relBuilder);
+
+        SupportsFilterPushDown.Result result = pushdownResultWithScan._1;
+        TableSourceTable tableSourceTable = pushdownResultWithScan._2;
+
+        FlinkLogicalTableSourceScan newScan =
+                FlinkLogicalTableSourceScan.create(scan.getCluster(), tableSourceTable);
+
+        // build new calc program
+        RexProgramBuilder programBuilder =
+                new RexProgramBuilder(newScan.getRowType(), call.builder().getRexBuilder());
+
+        if (!result.getRemainingFilters().isEmpty()) {
+            RexNode[] unconvertedPredicates = extractedPredicates._2;
+            RexNode remainingCondition =
+                    getRemainingConditions(relBuilder, result, unconvertedPredicates);
+
+            programBuilder.addCondition(remainingCondition);
+        }
+
+        List<RexNode> projects =
+                originProgram.getProjectList().stream()
+                        .map(originProgram::expandLocalRef)
+                        .collect(Collectors.toList());
+        List<String> outputFieldNames = originProgram.getOutputRowType().getFieldNames();
+
+        for (int i = 0; i < projects.size(); i++) {
+            programBuilder.addProject(projects.get(i), outputFieldNames.get(i));
+        }
+
+        Calc newCalc = calc.copy(calc.getTraitSet(), newScan, programBuilder.getProgram());
+        call.transformTo(newCalc);
```

Comment:
If remaining filters is empty and `originProgram.getProjectList()` is empty, we don't need to build a new calc.

Reply:
Do we want to remove the calc completely in that case?

Reply:
Yes. In `PushFilterIntoTableSourceScanRule` we only keep the scan if the filter is empty. I think we should do as same.